File: android/app/src/main/java/com/oasth/widget/data/OasthApi.kt
package com.oasth.widget.data

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * OASTH API client using native HTTP with session credentials
 */
class OasthApi(private val sessionManager: SessionManager) {
    
    companion object {
        private const val BASE_URL = "https://telematics.oasth.gr/api/"
    }
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    
    /**
     * Get arrivals for a specific stop
     */
    suspend fun getArrivals(stopCode: String): List<BusArrival> = withContext(Dispatchers.IO) {
        val session = sessionManager.getSession()
        
        val request = Request.Builder()
            .url("${BASE_URL}?act=getStopArrivals&p1=$stopCode")
            .addHeader("User-Agent", "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36")
            .addHeader("Accept", "application/json, text/javascript, */*; q=0.01")
            .addHeader("X-Requested-With", "XMLHttpRequest")
            .addHeader("X-CSRF-Token", session.token)
            .addHeader("Cookie", "PHPSESSID=${session.phpSessionId}")
            .addHeader("Origin", "https://telematics.oasth.gr")
            .addHeader("Referer", "https://telematics.oasth.gr/en/")
            .get()
            .build()
        
        val response = client.newCall(request).execute()
        
        if (response.code == 401) {
            // Session expired, refresh and retry
            sessionManager.refreshSession()
            return@withContext getArrivals(stopCode)
        }
        
        val body = response.body?.string() ?: "[]"
        
        try {
            val type = object : TypeToken<List<BusArrival>>() {}.type
            gson.fromJson<List<BusArrival>>(body, type) ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Get all bus lines
     */
    suspend fun getLines(): List<BusLine> = withContext(Dispatchers.IO) {
        val session = sessionManager.getSession()
        
        val request = Request.Builder()
            .url("${BASE_URL}?act=webGetLines")
            .addHeader("User-Agent", "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36")
            .addHeader("Accept", "application/json, text/javascript, */*; q=0.01")
            .addHeader("X-Requested-With", "XMLHttpRequest")
            .addHeader("X-CSRF-Token", session.token)
            .addHeader("Cookie", "PHPSESSID=${session.phpSessionId}")
            .post(okhttp3.RequestBody.create(null, ByteArray(0)))
            .build()
        
        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: "[]"
        
        try {
            val type = object : TypeToken<List<BusLine>>() {}.type
            gson.fromJson<List<BusLine>>(body, type) ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }
}

File: ios_app/lib/services/oasth_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';
import 'session_manager.dart';

/// OASTH API client using session credentials
class OasthApi {
  static const String _baseUrl = 'https://telematics.oasth.gr';
  static const String _apiUrl = '$_baseUrl/api/';
  
  final SessionManager _sessionManager;
  bool _isRetrying = false;
  
  OasthApi(this._sessionManager);
  
  /// Get arrivals for a specific stop
  Future<List<BusArrival>> getArrivals(String stopCode) async {
    try {
      final session = await _sessionManager.getSession();
      
      final response = await http.post(
        Uri.parse('$_apiUrl?act=getStopArrivals&p1=$stopCode'),
        headers: {
          'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15',
          'Accept': 'application/json, text/javascript, */*; q=0.01',
          'X-Requested-With': 'XMLHttpRequest',
          'X-CSRF-Token': session.token,
          'Cookie': 'PHPSESSID=${session.phpSessionId}',
          'Origin': _baseUrl,
          'Referer': '$_baseUrl/',
        },
      );
      
      // Check for unauthorized
      if (response.statusCode == 401 || 
          response.body.toLowerCase().contains('unauthorized') ||
          response.body.toLowerCase().contains('not authorized')) {
        if (!_isRetrying) {
          _isRetrying = true;
          await _sessionManager.refreshSession();
          final result = await getArrivals(stopCode);
          _isRetrying = false;
          return result;
        } else {
          _isRetrying = false;
          return [];
        }
      }
      
      try {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((item) => BusArrival.fromJson(item)).toList();
      } catch (e) {
        return [];
      }
    } catch (e) {
      return [];
    }
  }
  
  /// Get stop info by code
  Future<String?> getStopInfo(String stopCode) async {
    try {
      final session = await _sessionManager.getSession();
      
      final response = await http.post(
        Uri.parse('$_apiUrl?act=getStopArrivals&p1=$stopCode'),
        headers: {
          'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X)',
          'Accept': 'application/json',
          'X-Requested-With': 'XMLHttpRequest',
          'X-CSRF-Token': session.token,
          'Cookie': 'PHPSESSID=${session.phpSessionId}',
        },
      );
      
      // Try to extract stop description
      final regex = RegExp(r'"bstop_descr"\s*:\s*"([^"]+)"');
      final match = regex.firstMatch(response.body);
      return match?.group(1);
    } catch (e) {
      return null;
    }
  }
}
